jupyterhub/tests/test_roles.py

```diff
@@ -1114,3 +952,205 @@ async def test_user_group_roles(app, create_temp_role):
     assert len(reply[0]['roles']) == 1
     assert reply[0]['name'] == 'jack'
     assert group_role.name not in reply[0]['roles']
+
+
+async def test_config_role_list():
+    roles_to_load = [
+        {
+            'name': 'elephant',
+            'description': 'pacing about',
+            'scopes': ['read:hub'],
+        },
+        {
+            'name': 'tiger',
+            'description': 'pouncing stuff',
+            'scopes': ['shutdown'],
+        },
+    ]
+    hub = MockHub(load_roles=roles_to_load)
+    hub.init_db()
+    hub.authenticator.admin_users = ['admin']
+    await hub.init_role_creation()
+    for role_conf in roles_to_load:
+        assert orm.Role.find(hub.db, name=role_conf['name'])
+    # Now remove elephant from config and see if it is removed from database
+    roles_to_load.pop(0)
+    hub.load_roles = roles_to_load
+    await hub.init_role_creation()
+    assert orm.Role.find(hub.db, name='tiger')
+    assert not orm.Role.find(hub.db, name='elephant')
+
+
+async def test_config_role_users():
+    role_name = 'painter'
+    user_name = 'benny'
+    user_names = ['agnetha', 'bjorn', 'anni-frid', user_name]
+    roles_to_load = [
+        {
+            'name': role_name,
+            'description': 'painting with colors',
+            'scopes': ['users', 'groups'],
+            'users': user_names,
+        },
+    ]
+    hub = MockHub(load_roles=roles_to_load)
+    hub.init_db()
+    hub.authenticator.admin_users = ['admin']
+    hub.authenticator.allowed_users = user_names
+    await hub.init_role_creation()
+    await hub.init_users()
+    await hub.init_role_assignment()
+    user = orm.User.find(hub.db, name=user_name)
+    role = orm.Role.find(hub.db, name=role_name)
+    assert role in user.roles
+    # Now reload and see if user is removed from role list
+    roles_to_load[0]['users'].remove(user_name)
+    hub.load_roles = roles_to_load
+    await hub.init_role_creation()
+    await hub.init_users()
+    await hub.init_role_assignment()
+    user = orm.User.find(hub.db, name=user_name)
+    role = orm.Role.find(hub.db, name=role_name)
+    assert role not in user.roles
+
+
+async def test_duplicate_role_users():
+    role_name = 'painter'
+    user_name = 'benny'
+    user_names = ['agnetha', 'bjorn', 'anni-frid', user_name]
+    roles_to_load = [
+        {
+            'name': role_name,
+            'description': 'painting with colors',
+            'scopes': ['users', 'groups'],
+            'users': user_names,
+        },
+        {
+            'name': role_name,
+            'description': 'painting with colors',
+            'scopes': ['users', 'groups'],
+            'users': user_names,
+        },
+    ]
+    hub = MockHub(load_roles=roles_to_load)
+    hub.init_db()
+    with pytest.raises(AttributeError):
```

Comment:
```suggestion
    with pytest.raises(ValueError):
```

Reply:
Should be raising `ValueError` from `app.py`, does it?

Reply:
Indeed, thanks for catching that